docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+In order to control whether this output is used or not, a new entry is added to the `IncrementalGeneratorOutputKind` enum:
+
+```diff
+[Flags]
+public enum IncrementalGeneratorOutputKind
+{
+    None = 0,
+    Source = 0b1,
+    PostInit = 0b10,
++    Artifact = 0b100
+}
+```
+
+A host such as the IDE may disable non-source outputs when constructing the Generator Driver.
+
+In terms of the compilation and inputs that an Artifact can be based on, artifact outputs are treated the same as source outputs; all generation is based on the state of the compilation *before* any other generation has happened, and an artifact can't use the output of another generator (outside of itself) to generate artifacts.
+
+### File names and output locations
+
+The call to `AddFile` takes a `hintName` for the output file rather than an explicit file name or path. This allows the host of the generator driver the freedom to decide where precisely the file should go (if it is indeed going to e.g. the disk). Here we propose a set of rules that the command line compiler will follow. Other hosts may choose a different scheme.
+
+The compiler will take a new command line option `/generatedartifactsout:` that will specify the _default_ location of any generated artifacts. By default, using the compiler MSBuild targets, this will be set to the `$(OutDir)` property (i.e. `/bin/`). Note that this is different to `OutputAssembly` and `GeneratedFilesOutputPath` which typically use the `$(IntermediateOutputPath)` (i.e. `/obj/`). This ensures that artifacts are written to their final location a single time, rather than being written to an intermediate directory and having to be copied, optimizing the disk usage.
+
+By default the supplied `hintName` will be used as the filename. We already have conditions in place for source texts to prevent hint names from being invalid filenames, and we will reuse this logic for artifact names. If two generators attempt to write a file with the same name, the generator name, including namespace, will be prepended to the name of the files to ensure there are no conflicts.
+
+### Customizing output locations
+
+While it may be acceptable for the majority of cases to simply place the artifacts in the bin directory, there are known cases (such as the Javascript interop case mentioned at the beginning) where it would be beneficial to place generated files in a more specific location (such as `wwwroot`). This proposal provides a mechanism to override the location of artifacts in two ways: on a per-generator basis, and on an individual artifact basis. Both of these mechanisms are implemented via `.globalconfig` options, and exposed with `MSBuild` properties to enable easy configuration by both Generator authors and consumers.
+
+The generator driver will recognize the following `.globalconfig` options:
+
+- `compiler_artifact.<generatorname>.output_dir = <dir>`
+- `compiler_artifact.<artifactname>.output_dir = <dir>`
```

Comment:
This feels really disconnected. What do we gain by the .globalconfig route? 

Lots of ways to go wrong - typos, assuming a default is in use when an entry exists, changing an artifact name and missing the update to this file...

I think folks are more likely to think "here's my file content, the desired name, and where it will go"